### DISCUSSION TITLE
Can login UI be a SPA with it's own backend?

### DISCUSSION THREAD

**Discussion by alekbarszczewski** [ory/hydra#2951] [Q&A] [answered]
With Ory Hydra I need my own login UI to which Hydra will redirect user to login. This UI has to be server-side app that will call Hydra admin api to accept/reject login when user clicks submit button. I am wondering if it's ok for this UI to be SPA with it's own backend and API instead of server-side rendered HTML? The problem is that we want to make it look same as rest of the app which is a SPA, and on login page currently it uses some JS dynamic field validation, some dynamic menu, etc. So is it okay to use one of following approaches or maybe it is not recommended for some reasons?:

![Hydra@2x](https://user-images.githubusercontent.com/2742373/150551893-2bc817bb-421c-45fc-8680-d6cf493d0d24.png)

The reason is that for frontend team it would be much easier to implement login page(s) using approach (SPA + API) we currently use in all other places.

**Comment by grantzvolsky** [comment, marked as answer]
Hi, the answer is yes, the login/consent app can be a SPA with an API backend. The Hydra flows, however, involve redirects, so the user-agent will end up loading the SPA multiple times. Approach 1 above is plausible.
